RecoHGCal/GraphReco/plugins/peprCandidateFromHitProducer.cc

```diff
@@ -0,0 +1,345 @@
+
+        }
+    //}
+
+
+
+    event.put(std::move(candidates));
+
+    std::cout << "Results produced and put in event" << std::endl;
+
+
+    // clear all windows
+    for (auto& window : windows_) {
+        window.clear();
+    }
+
+    std::cout << "Windows cleared" << std::endl;
+}
+
+
+
+void peprCandidateFromHitProducer::fillWindows(const edm::Event& event) {
+
+    if (!windows_.size()) {
+        throw cms::Exception("NoWindows") << "no windows initialized";
+    }
+    
+    std::cout << "Number of windows = " << windows_.size() << std::endl;
+    
+    // copied block from window ntupler code
+    // get rechits, get positions and merge collections
+    std::vector<HGCRecHitWithPos> allrechits;
+    for (auto & token : recHitTokens_) {
+       for (const auto& rh : event.get(token)) {
+           HGCRecHitWithPos rhp = { const_cast<HGCRecHit*>(&rh), recHitTools_.getPosition(rh.detid()) };
+           allrechits.push_back(rhp);
+       }
+    }
+    // sort according to the energy
+    std::sort(allrechits.begin(), allrechits.end(), 
+        [](const HGCRecHitWithPos& rh1, const HGCRecHitWithPos& rh2) 
+            { return rh1.hit->energy() > rh2.hit->energy();});
+
+    
+    // fills a vector of the specified size with zeroes (entries will be 0 if rechit is not filled, and 1 if it is filled)
+    std::vector<size_t> filledrechits(allrechits.size(),0);
+
+
+    for (auto & window : windows_) {
+        //fill rechits in this window
+        for(size_t it=0;it<allrechits.size();it++) {
+           if(filledrechits.at(it)>3) continue;
+           if(window.maybeAddRecHit(allrechits.at(it)))
+               filledrechits.at(it)++;
+        }
+
+        window.fillFeatureArrays();
+    }
+
+
+}
+
+
+void peprCandidateFromHitProducer::writeInputArrays(const std::vector<std::vector<float> >& hitFeatures) {
+
+    std::string inpipeRAM = "/dev/shm/" + inpipeName_; //write in RAM
+    //std::ofstream inputArrayStream(inpipeName_.c_str()); //do not write in RAM (local testing)
+    std::ofstream inputArrayStream(inpipeRAM.c_str()); 
+
+    for (size_t i=0; i<hitFeatures.size(); i++) {
+
+        for (size_t j=0; j<hitFeatures[i].size(); j++) {
+            inputArrayStream << hitFeatures[i][j] << " ";
+        }
+        inputArrayStream << "\n";
+    }
+
+    inputArrayStream.close();
+}
+
+
+void peprCandidateFromHitProducer::readOutputArrays() {
+
+    //FIXME: currently just reading and printing, proper values to be stored
+    //FIXME: read until you get a last empty line, because the fifo is being filled not instantly
+
+    std::string outpipeRAM = "/dev/shm/" + outpipeName_; //read from RAM
+    //std::ifstream outputArrayStream(outpipeName_.c_str()); //do not read from RAM (local testing)
+    std::ifstream outputArrayStream(outpipeRAM.c_str()); 
+
+    std::string line;
+    if(outputArrayStream.is_open()) {
+        std::cout << "Output array file opened!" << std::endl;
+        while ( getline (outputArrayStream,line) )
```

Comment:
This needs to be:
while(true){
getline(...)
if(line is empty){break; }
}

Then it is waiting for all the output to be transferred before it stops reading